### PR TITLE
feat(logger): add a simple logging module

### PR DESCRIPTION
**Pull request by ullasholla**
This PR adds a simple logging module. It uses [Winston](https://github.com/winstonjs/winston) to do the heavy-lifitng but the API is agnostic to Winston. This is done so that Winston could be replaced by another logging module in the future if desired. The clients of the library will continue to function without any changes. The [README](https://github.com/SAP/open-ux-tools/blob/feat/add-logger/packages/logger/README.md) document has more information about the module.

### CODE REVIEW

**Comment by ullasholla** [comment]
Test coverage:
image

**Review by tobiasqueck — APPROVED** [review]
* well structured and easy to understand coding
* simple to use but also very flexible if specific configurations are needed
* I especially like the dev-null logging :)
* tried it for my use cases in the `axios-extension` and it worked without problems
* minor suggestion (but not required): user type `Partial<LoggerOptions> for the logger options property of the constructor so that I can also just provide a log level and not a transport (or vice versa)
* please add more js-doc descriptions to your functions

**Comment by tobiasqueck** [comment]
@zdravko-georgiev can you check if this works with logging in ui5 middlewares and tasks?

**Comment by zdravko-georgiev** [comment]
> @zdravko-georgiev can you check if this works with logging in ui5 middlewares and tasks?

@tobiasqueck Checking

**Comment by zdravko-georgiev** [comment]
Is this how the `info` supposed to look like in the console?
image (8)

**Comment by ullasholla** [comment]
> Is this how the `info` supposed to look like in the console? image (8)

Yes. I can't fault you for not getting all way the way through the [README](https://github.com/SAP/open-ux-tools/blob/feat/add-logger/packages/logger/README.md#not-supported) :P

image

**Comment by ullasholla** [comment]
@IainSAP, @zdravko-georgiev, @tobiasqueck: I changed the default format for console to be plain text instead of JSON (it is colored and timestamped though. Piping the output will not use the [color encoding](https://en.wikipedia.org/wiki/ANSI_escape_code) - this is to facilitate both manual inspection and text manipulation (color encoding can be distracting and painful to work with)). `FileTransport` still uses JSON.

@IainSAP, I am slightly deviating from the format used in Fiori tools (levels are not uppercased and the log entry is not enclosed by []). Let me know if you have any concerns with these. (Not using uppercase as I use `format.colorize()` and it expects lowercased levels. The effort to keep both didn't seem worth it at this point).

An example logger that logs to console and two files simultaneously (the transports are added at different points in time and don't contain identical logs/levels):
image


Piping the output to a file (`sample.log` does not have any ANSI escape codes):
image

@zdravko-georgiev, looks like the ui5-logger uses [`npmlog`](https://github.com/npm/npmlog) for logging. I'm still checking to see if the output can be made compatible.

File: packages/logger/src/winston-logger/vscode-output-channel-transport.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import Transport from 'winston-transport';
+// import { window, OutputChannel } from 'vscode';
```
Comment (zdravko-georgiev) [inline packages/logger/src/winston-logger/vscode-output-channel-transport.ts:2]: Can we remove this?

**Comment by zdravko-georgiev** [comment]
There is one empty file `packages/logger/foo`. Can we remove it?

**Comment by IainSAP** [comment]
Hi Ullas,

How will this interact with the existing SAP vscode logging we use in App Gen. Currently we already have rolling file and channel logging provided by the SAP VSCode logger. We do not want multiple log files being created when generating apps for example, or have multiple output channels for different parts of generation. This would complicate the logging. Does the open source logging proposal support integration with the existing SAP VSCode logging being used?

-Iain


From: Ullas Holla ***@***.***>
Reply to: SAP/open-ux-tools ***@***.***>
Date: Tuesday 11 January 2022 at 20:17
To: SAP/open-ux-tools ***@***.***>
Cc: Iain O'Farrell ***@***.***>, Mention ***@***.***>
Subject: Re: [SAP/open-ux-tools] feat(logger): add a simple logging module (PR #257)


@IainSAP<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FIainSAP&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=PbRMLerJlgLRVaoE4P6%2FUG4CeewdHsnU49YolkiAvBg%3D&reserved=0>, @zdravko-georgiev<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fzdravko-georgiev&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=za%2Fiy4sjFZPlTf9zSUZZBUW1TTF2MhnmZYF9w4XTvIU%3D&reserved=0>, @tobiasqueck<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Ftobiasqueck&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=4DWkQFLHsWxH76ztgZu2uA5L11Ii1OBF4ycLKphCLMY%3D&reserved=0>: I changed the default format for console to be plain text instead of JSON (it is colored and timestamped though. Piping the output will not use the color encoding<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fen.wikipedia.org%2Fwiki%2FANSI_escape_code&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=Hc1%2FOSOgkCzs42b0qr1BomXz2YHxcpsndvU%2BbJX3e5c%3D&reserved=0> - this is to facilitate both manual inspection and text manipulation (color encoding can be distracting and painful to work with)). FileTransport still uses JSON.

@IainSAP<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FIainSAP&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=PbRMLerJlgLRVaoE4P6%2FUG4CeewdHsnU49YolkiAvBg%3D&reserved=0>, I am slightly deviating from the format used in Fiori tools (levels are not uppercased and the log entry is not enclosed by []). Let me know if you have any concerns with these. (Not using uppercase as I use format.colorize() and it expects lowercased levels. The effort to keep both didn't seem worth it at this point).

An example logger that logs to console and two files simultaneously (the transports are added at different points in time and don't contain identical logs/levels):
[image]<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F212316%2F149013659-06005d72-cc29-455b-b7ca-071d5a1aa1d4.png&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=XqjbNlNGGzNNA2UW%2BbokgXw7P%2B33UQl9Q3OjBqCIxt0%3D&reserved=0>

Piping the output to a file (sample.log does not have any ANSI escape codes):
[image]<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F212316%2F149013774-959479b4-164d-45bb-ab89-bb02d55367d7.png&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=M99YaXmwUv1bi4z2GhD7lpaLhZOc%2FerCNv0UXVpTE0Q%3D&reserved=0>

@zdravko-georgiev<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fzdravko-georgiev&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=za%2Fiy4sjFZPlTf9zSUZZBUW1TTF2MhnmZYF9w4XTvIU%3D&reserved=0>, looks like the ui5-logger uses npmlog<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnpm%2Fnpmlog&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=fTXUmm02bz56fHh%2FJXAtxEu%2FTLwORxRPA6NKOcyVhEw%3D&reserved=0> for logging. I'm still checking to see if the output can be made compatible.

—
Reply to this email directly, view it on GitHub<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FSAP%2Fopen-ux-tools%2Fpull%2F257%23issuecomment-1010326678&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=cSDpbMtkvP0H7fUP1jvlxuv92sHRiRLARvfwHUAStSA%3D&reserved=0>, or unsubscribe<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FALDBLRWJAVGNG4GW5HOECKDUVSF53ANCNFSM5LVQM5WQ&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=hEuiSFlQ7FtBUWvLuMmj2F2zqHm6frjgvyi0mM2zMnI%3D&reserved=0>.
Triage notifications on the go with GitHub Mobile for iOS<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fapps.apple.com%2Fapp%2Fapple-store%2Fid1477376905%3Fct%3Dnotification-email%26mt%3D8%26pt%3D524675&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=Z9194aBDBMEVyghyWupY0svEK51MnaCnr76v3dFzaX4%3D&reserved=0> or Android<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dcom.github.android%26referrer%3Dutm_campaign%253Dnotification-email%2526utm_medium%253Demail%2526utm_source%253Dgithub&data=04%7C01%7Ciain.o-farrell%40sap.com%7C0e8c2e0e431b40b7d56208d9d53f67cd%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637775290566087863%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=Q3p6THjJKamHwFeDEGxRvxgyJd%2F1YAOqG36h%2Fdl7JCo%3D&reserved=0>.

**Comment by ullasholla** [comment]
> There is one empty file `packages/logger/foo`. Can we remove it?

That's there for good luck 😬

File: packages/logger/src/winston-logger/logger.ts [hunk 2]
```diff
@@ -39,6 +41,18 @@ export class WinstonLogger implements Logger {
...
     }
+    log(data: string | Log): void {
+        if (!this.transportMap.size) {
+            // Nothing to do
+            return;
+        }
+        if (typeof data === 'string') {
+            this._logger.log(this.winstonLevel, data);
+        } else {
+            this._logger.log(toWinstonLogLevel(data.level)!, data.message);
```
Comment (zdravko-georgiev) [inline packages/logger/src/winston-logger/logger.ts:52]: Is the `!` on purpose?

Comment (ullasholla) [reply packages/logger/src/winston-logger/logger.ts]: Yes, it was on purpose but it's gone now.

**Comment by ullasholla** [comment]
Child loggers are now supported. They add an prefix to the log to make it clear which component is doing the logging. The `FileTransport` will have the property `label` to filter the logs on, if required.

On the console, using child loggers looks like this:
image

**Review by zdravko-georgiev — APPROVED** [review]
Code looks good, 2 Unit test are failing

**Review by tobiasqueck — APPROVED** [review]
Reviewed:

- [x] understandability of the code - nice, clean structure, however a little more complex than I expected 
- [x] very good test coverage
- [x] works with planned `axios-extension` module

**Comment by tobiasqueck** [comment]
> Code looks good, 2 Unit test are failing

@zdravko-georgiev: not for me, which ones cause problems for you?

**Comment by zdravko-georgiev** [comment]
> Details

@tobiasqueck The PR build is failing https://github.com/SAP/open-ux-tools/runs/4813000583?check_suite_focus=true

File: package.json [hunk 3]
```diff
@@ -6,6 +6,7 @@
     "devDependencies": {
         "@changesets/cli": "2.19.0",
         "@types/jest": "27.4.0",
+        "@types/node": "17.0.8",
```
Comment (devinea) [inline package.json:9]: Should opt for the lowest supported version of node (e.g. 12) and only LTS versions

File: packages/logger/.eslintrc [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+        "jsdoc/require-jsdoc": 0,
+        "jsdoc/match-description": 0,
+        "jsdoc/multiline-blocks": 0,
+        "jsdoc/newline-after-description": 0,
+        "jsdoc/require-param-description": 0,
+        "jsdoc/no-multi-asterisks": 0,
+        "jsdoc/check-tag-names": 0,
+        "@typescript-eslint/no-unused-vars": ["error", { "varsIgnorePattern": "^_", "argsIgnorePattern": "^_" }]
+    }
+}
```
Comment (devinea) [inline packages/logger/.eslintrc:16]: why have custom rules here. Shouldn't all modules use the same?

File: packages/logger/README.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+You can `add()`, `remove()` transport dynamically. `transports()` returns a list of the current transports of the logger.
+
+### Examples
+The unit test file can be used as an example to explore the logger's API: [./test/unit/wiston-logger/logger.test.ts](./test/unit/wiston-logger/logger.test.ts)
+
+Please refer to the debugging section in the project root [README](../../README.md) for help debugging the test.
+### Not supported
+The following options are not supported yet. There are no definite plans to add them in the future. They will be considered on a need-basis.
+
+* Child loggers
```
Comment (devinea) [inline packages/logger/README.md:70]: This has been added since?

File: packages/logger/package.json [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    "main": "dist/index.js",
+    "scripts": {
+        "build": "pnpm clean && tsc",
+        "watch": "tsc --watch",
+        "clean": "rimraf dist coverage",
+        "format": "prettier --write '**/*.{js,json,ts,yaml,yml}' --ignore-path ../../.prettierignore",
+        "lint": "eslint . --ext .ts",
+        "lint:fix": "eslint . --ext .ts --fix",
+        "test": "jest --maxWorkers=1 --colors",
+        "test-u": "jest --maxWorkers=1 --colors -u",
```
Comment (devinea) [inline packages/logger/package.json:16]: I recall you asked to remove `test-u` from the other modules

File: packages/logger/package.json [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        "link": "pnpm link --global",
+        "unlink": "pnpm unlink --global"
+    },
+    "dependencies": {
+        "chalk": "4.1.2",
+        "lodash.clonedeep": "4.5.0",
```
Comment (devinea) [inline packages/logger/package.json:22]:
```suggestion
        "lodash": "4.17.21"
```
please use `lodash` instead and `import cloneDeep from 'lodash/clonedeep';`
Makes for better treeshaking with webpack 🤷

File: packages/logger/package.json [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        "logform": "2.3.2",
+        "winston": "3.3.3",
+        "winston-transport": "4.4.1"
+    },
+    "devDependencies": {
+        "@types/debug": "4.1.7",
+        "@types/lodash.clonedeep": "4.5.6",
```
Comment (devinea) [inline packages/logger/package.json:29]:
```suggestion
        "@types/lodash": "4.14.176"
```

File: packages/logger/src/types.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
+import cloneDeep from 'lodash.clonedeep';
```
Comment (devinea) [inline packages/logger/src/types.ts:1]:
```suggestion
import cloneDeep from 'lodash/clonedeep';
```

File: packages/logger/tsconfig.json [hunk 10]
```diff
@@ -0,0 +1,12 @@
+{
+    "include": ["src/**/*.ts"],
+    "compilerOptions": {
+        "declaration": true,
+        "outDir": "dist",
+        "sourceMap": true,
+        "baseUrl": "./src",
+        "resolveJsonModule": true,
+        "lib": ["ESNext"]
```
Comment (devinea) [inline packages/logger/tsconfig.json:9]:
```suggestion
        "lib": ["2019"]
```
safer for node 12 support.

File: packages/logger/src/winston-logger/logger.ts [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+class BaseWinstonLogger implements Logger {
+    protected _logger: winston.Logger;
+    protected logPrefix: string;
+    protected logPrefixColor: string;
+    protected winstonLevel: string;
+    protected metadataOverride?: Metadata;
+    // Maintain of map of transports. This is useful for adding/removing transports
+    protected transportMap: Map<Transport, WinstonTransport>;
+    // Cache child loggers by logPrefix. Use weak references to avoid causing memory leaks
+    protected childMap: Map<string, WeakRef<Logger>> = new Map();
```
Comment (devinea) [inline packages/logger/src/winston-logger/logger.ts:34]: It seems WeakRef is only supported from Node 14.6.0+. We need node 12 support for a few more months atleast.

**Review by devinea — CHANGES_REQUESTED** [review]

File: packages/logger/package.json (hunk 6 above)
Comment (ullasholla) [reply packages/logger/package.json:16]: I kept this to be consistent with the other modules. Shall I remove this from all the modules?

**Comment by ullasholla** [comment]
> @tobiasqueck The PR build is failing https://github.com/SAP/open-ux-tools/runs/4813000583?check_suite_focus=true

@zdravko-georgiev & @tobiasqueck, this is to do with Node 12 vs Node 14. Like @devinea pointed out, it's to do with `WeakRef` not being supported yet on Node 12 yet.

`WeakRef`s are used to keep to make sure the cache of child loggers does not prevent the garbage collection of the loggers. If we don't use weak references, there's a small chance of memory leak (we can remove support for weak refs now and add it back when we support only Node 14+). Another option is not to cache them in the first place.

Thoughts?

File: packages/logger/package.json (hunk 6 above)
Comment (devinea) [reply packages/logger/package.json:16]: may as well

File: packages/logger/README.md (hunk 5 above)
Comment (ullasholla) [reply packages/logger/README.md:70]: Good catch, thanks.

File: packages/logger/package.json (hunk 6 above)
Comment (ullasholla) [reply packages/logger/package.json]: Changed my mind, don't want to make changes in the other modules here. Removed in logger, ta.

File: packages/logger/README.md (hunk 5 above)
Comment (ullasholla) [reply packages/logger/README.md]: Fixed.

File: packages/logger/package.json (hunk 7 above)
Comment (ullasholla) [reply packages/logger/package.json]: Fixed in https://github.com/SAP/open-ux-tools/pull/257/commits/34fab90d6a1a69d58b44c43d0be32b66836d41df

File: packages/logger/src/types.ts (hunk 9 above)
Comment (ullasholla) [reply packages/logger/src/types.ts]: Fixed in https://github.com/SAP/open-ux-tools/pull/257/commits/34fab90d6a1a69d58b44c43d0be32b66836d41df

File: packages/logger/src/winston-logger/logger.ts (hunk 11 above)
Comment (ullasholla) [reply packages/logger/src/winston-logger/logger.ts]: Removed the need for `WeakRef`. Fixed in https://github.com/SAP/open-ux-tools/pull/257/commits/c2348bd8d23c4b47495add3e57d3881c8eb5241f

File: packages/logger/tsconfig.json (hunk 10 above)
Comment (ullasholla) [reply packages/logger/tsconfig.json]: Fixed in https://github.com/SAP/open-ux-tools/pull/257/commits/c2348bd8d23c4b47495add3e57d3881c8eb5241f

File: packages/logger/package.json (hunk 8 above)
Comment (ullasholla) [reply packages/logger/package.json]: Fixed in https://github.com/SAP/open-ux-tools/pull/257/commits/34fab90d6a1a69d58b44c43d0be32b66836d41df

File: packages/logger/.eslintrc (hunk 4 above)
Comment (ullasholla) [reply packages/logger/.eslintrc:15]: Fixed in https://github.com/SAP/open-ux-tools/pull/257/commits/0ad205679050af6a49ee1e7619f136df131219a8

File: package.json (hunk 3 above)
Comment (ullasholla) [reply package.json]: Fixed in https://github.com/SAP/open-ux-tools/pull/257/commits/c2348bd8d23c4b47495add3e57d3881c8eb5241f

**Comment by ullasholla** [comment]
> Reviewed:
> 
>     * [x]  understandability of the code - nice, clean structure, however a little more complex than I expected

Thanks @tobiasqueck. Some of the complexity was due to caching. I've removed it. The caching covered a niche usecase and I think the tradeoff of having to worry about memory leaks (and having to use weak references is not worth it). Let me know if there are other areas where the code seems more complicated than necessary. We can think of refactoring those areas.

The PR tests now pass (they failed because they run on Node 12 but I developed locally using Node 14. I've removed code that doesn't work with Node 12 and have switched to using Node 12 locally as well).

I have addressed @devinea's comments (I have not marked all the conversations as resolved. I started doing that and then realized that it may be easier for to track the changes if I added a comment but left them as-is. Let me know what works better and I'll keep that in mind for the future).

**Review by tobiasqueck — APPROVED** [review]

**Review by devinea — DISMISSED** [review]
thanks @ullasholla

**Review by devinea — APPROVED** [review]

**Review by tobiasqueck — APPROVED** [review]

**Pull request merged**